processor.go

```diff
@@ -96,204 +116,193 @@ func (p *Processor[AC, OC, JC]) init() {
 	}
 	// Make a map of triggers to states so we can easily reference it
 	p.stateMap = map[string]State[AC, OC, JC]{}
-	for _, s := range p.states {
-		p.stateMap[s.TriggerState] = s
-	}
-	// get a list of return state names for use
-	p.stateNames = make([]string, 0, len(p.stateMap))
-	for k := range p.stateMap {
-		p.stateNames = append(p.stateNames, k)
-	}
-
-	// For each state, we need a channel of jobs
+	p.stateStatusMap = map[string]*StatusCount{}
+	p.stateWaitingJobsMap = map[string][]Job[JC]{}
 	p.stateChan = map[string]chan Job[JC]{}
+	p.sortedStateNames = []string{}
 
-	// Create the state chans
-	totalConcurrency := 0
 	for _, s := range p.states {
-		if s.Terminal {
-			continue
+		stateName := s.TriggerState
+
+		p.sortedStateNames = append(p.sortedStateNames, stateName)
+		p.stateMap[stateName] = s
+		p.stateStatusMap[stateName] = &StatusCount{
+			State:    stateName,
+			Terminal: s.Terminal,
 		}
-		p.stateChan[s.TriggerState] = make(chan Job[JC], s.Concurrency) // make a chan
-		totalConcurrency += s.Concurrency
+		// This is by-design unbuffered
+		p.stateChan[stateName] = make(chan Job[JC])
 	}
 
-	// When a job changes state, we send it to this channel to centrally manage and re-queue
-	p.returnChan = make(chan Return[JC], totalConcurrency*2) // make it the size of the total amount of in flight jobs we could have so that each worker can return a task
+	sort.Strings(p.sortedStateNames)
+
+	// This is by-design unbuffered
+	p.returnChan = make(chan Return[JC])
 }
 
 // Exec this big work function, this does all the crunching
 func (p *Processor[AC, OC, JC]) Exec(ctx context.Context, r *Run[OC, JC]) error {
+	if err := p.validate(p.states); err != nil {
+		return err
+	}
+
 	p.init()
 
 	if p.allJobsAreTerminal(r) {
 		slog.Info("AllJobsTerminal")
 		return nil
 	}
 
-	wg := sync.WaitGroup{}
-
 	// create the workers
 	for _, s := range p.states {
 		// Terminal states don't need to recieve jobs, they're just done
 		if s.Terminal {
 			continue
 		}
-		if s.Exec == nil {
-			return p.invalidStateError(s.TriggerState)
-		}
-
-		concurrency := s.Concurrency
 
 		// Make workers for each, they just process and fire back to the central channel
-		for i := 0; i < concurrency; i++ { // add a waiter for every go processor, do it before forking
-			labels := pprof.Labels("type", "jorbWorker", "state", s.TriggerState, "id", fmt.Sprintf("%d", i))
-			pprof.Do(ctx, labels, func(_ context.Context) {
-				go p.execFunc(ctx, r, s, i, wg)()
+		for i := 0; i < s.Concurrency; i++ {
+			p.wg.Add(1)
```

Comment:
This whole for loop may make more sense in StateExec

Reply:
fixed in v2